### PR TITLE
Jmcda001/day2/puzzles

### PR DESCRIPTION
**Pull request by jmcda001**

### CODE REVIEW

File: days/src/day2/Day2.kt [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            .filter { it }.size
+
+    companion object {
+        fun buildFromFile(filename: String): Day2 {
+            val policyPasswordRegex = "(\\d+)-(\\d+) (\\w): (\\w+)".toRegex()
+            return Day2(File(filename).useLines { it.toList() }.asSequence()
+                    .map { policyPasswordRegex.matchEntire(it) }
+                    .filterNotNull()
+                    .map { matchResult ->  matchResult.groupValues.drop(1) }
+                    .map { Pair(Pair(Pair(it[0].toInt(), it[1].toInt()), it[2][0]), it[3])}
```
Comment (brrcrites) [inline days/src/day2/Day2.kt:38]: Might be cleaner to replace this `Pair(Pair(Pair()))` with a data class and then have the `PolicyPuzzle*` take that as the input

File: days/src/day2/Day2.kt [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+interface Policy {
+    fun isValid(password: String): Boolean
+}
+data class PolicyPuzzle1(val low: Int, val high: Int, val symbol: Char): Policy {
+    override fun isValid(password: String) = password.count { it == symbol } in low..high
+}
+
+data class PolicyPuzzle2(val index1: Int, val index2: Int, val symbol: Char): Policy {
+    override fun isValid(password: String) = (password[index1-1] == symbol || password[index2-1] == symbol) &&
+                    password[index1-1] != password[index2-1]
```
Comment (brrcrites) [inline days/src/day2/Day2.kt:15]: I truncated this to `((password[index1-1] == symbol) != (password[index2-1] == symbol))`, although it looks like Kotlin has [a logical XOR operator](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-boolean/xor.html)

**Review by brrcrites — APPROVED** [review]

File: days/src/day2/Day2.kt (hunk 1 above)
Comment (jmcda001) [reply days/src/day2/Day2.kt:38]: I had that before but I didn't like how the variables didn't match up on the second puzzle. I may try to figure enums into my solutions more (still a bit weak on those in Kotlin I think)

Comment (brrcrites) [reply days/src/day2/Day2.kt:38]: The cleanest and most extensible (but also most verbose) way would probably be to [setup an interface](https://kotlinlang.org/docs/reference/interfaces.html) and then created the two data classes with separate naming conventions that matched that interface?

**Pull request merged**